Basic flow, remove children on parent

cryogenian opened the pull request:
fixes #263 
fixes #175

jdegoes commented:
@garyb Can I get a review / merge?

garyb commented on src/Model/Notebook.purs:
```diff
@@ -21,6 +21,7 @@ type State =
   , tickDate :: Maybe Date
   , platform :: Platform
   , dialog :: Maybe Dialog
+  , requesting :: Maybe CellId
```
Is it right that we can only ever have one cell making a request at any time? Perhaps this should be a list instead?

Maybe it's fine, as obviously a user can't request content for multiple cells simultaneously, but is there never a case where we would need to do it either?

garyb commented on src/Driver/Notebook/Notify.purs:
```diff
@@ -0,0 +1,78 @@
+  notify notebook requestedId cellId nKnot driver 
+notifyChildren notebook nKnot (CellSlamDownEvent cellId _) driver requestedId =
+  notify notebook requestedId cellId nKnot driver
+notifyChildren _ _ _ _ _ = pure unit
```
Perhaps we should filter `input` in `notifyDriver` instead, as then we probably don't need `notifyChildren` at all.

garyb commented on src/Driver/Notebook/Notify.purs:
```diff
@@ -0,0 +1,78 @@
+import Utils (elem)
+
+
+type NotifyKnot = Maybe Timeout 
```
Why does the notify stuff run on a `Timeout`? Does this not present a problem if you run one cell with children and then another before the timeout executes? I'd think running the second cell would overwrite the pending notification request from the first cell.

garyb commented on src/Driver/Notebook/Notify.purs:
```diff
@@ -0,0 +1,78 @@
+    -1 -> for_ (dependentCells cellId) request
+    x -> maybe (for_ ds request) request 
+        ((ancestors <> [requestedId]) !! (x + 1) >>=
+         \x -> notebook ^? cellById x)
```
This seems really confusing, is there some way it could be simplified or perhaps broken up into things with meaningful names to help?

garyb commented:
This isn't a comment on this PR in particular, but a related question for @jdegoes: currently, if you create a query cell from the result of another, we have a dependency between the two. If the query is changed so that it no longer selects from the previous cell's output the dependency is still there in the graph - is that something we're going to have to handle at some point? (Same applies to some other cell types too I guess, but query is the trickier one as it doesn't have an explicit file selector or anything like search does, for example).

jdegoes commented:
>  If the query is changed so that it no longer selects from the previous cell's output, the dependency is still there in the graph - is that something we're going to have to handle at some point?

Yep, unfortunately. For now, it's OK that we don't update the dependency. Eventually, the graph must reflect the actual, current dependency between the nodes, regardless of how that has changed over time.

cryogenian commented:
Thanks for review! 
* changed `Maybe CellId` to `[CellId]` in `requesting` field of notebook state
* removed `notifyChildren`
* removed timeouts from `CellResult` case, left only in `CellSlamDownEvent`, changed `NotifyKnot` from `Maybe Timeout` to `Map CellId Timeout`
* reworked `notify`, added comments. Hope it's better now.

garyb commented:
Looks great, and the changes make it much clearer now. Good stuff! :+1:
